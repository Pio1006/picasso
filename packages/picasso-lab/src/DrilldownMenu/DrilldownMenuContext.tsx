import { createContext } from 'react'

export interface DrilldownMenuContextProps {
  activeMenuKey?: string
  setActiveMenuKey?: (key?: string) => void
}

export default createContext<DrilldownMenuContextProps>({})
