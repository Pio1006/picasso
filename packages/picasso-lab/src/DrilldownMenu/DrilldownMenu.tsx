import { BaseProps, CompoundedComponentWithRef } from '@toptal/picasso-shared'
import MenuList, { MenuListAttributes } from '@toptal/picasso/MenuList'
import React, { forwardRef, ReactNode, useMemo, useState } from 'react'

import DrilldownMenuItem from '../DrilldownMenuItem'
import DrilldownMenuContext from './DrilldownMenuContext'

export interface Props extends BaseProps, MenuListAttributes {
  /* Menu items */
  children?: ReactNode
}

export interface StaticProps {
  Item: typeof DrilldownMenuItem
}

export const DrilldownMenu = forwardRef<HTMLUListElement, Props>(
  function Drilldown (props, ref) {
    const { className, style, children, ...rest } = props
    const [activeMenuKey, setActiveMenuKey] = useState<string>()
    const context = useMemo(() => ({ activeMenuKey, setActiveMenuKey }), [
      activeMenuKey
    ])

    return (
      <DrilldownMenuContext.Provider value={context}>
        <MenuList {...rest} ref={ref} className={className} style={style}>
          {children}
        </MenuList>
      </DrilldownMenuContext.Provider>
    )
  }
) as CompoundedComponentWithRef<Props, HTMLUListElement, StaticProps>

DrilldownMenu.Item = DrilldownMenuItem

export default DrilldownMenu
