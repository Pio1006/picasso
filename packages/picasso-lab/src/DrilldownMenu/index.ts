import { OmitInternalProps } from '@toptal/picasso-shared'

import { Props } from './DrilldownMenu'

export { default } from './DrilldownMenu'
export type DrilldownMenuProps = OmitInternalProps<Props>
