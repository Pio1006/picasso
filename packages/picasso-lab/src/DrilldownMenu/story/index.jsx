import menuItemStory from '../../DrilldownMenuItem/story'
import { DrilldownMenu } from '../DrilldownMenu'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Picasso Lab').createPage(
  'DrilldownMenu',
  `
    Nested menu list.
  `
)

page
  .createTabChapter('Props')
  .addComponentDocs({ component: DrilldownMenu, name: 'Drilldown' })
  .addComponentDocs(menuItemStory.componentDocs)

page
  .createChapter()
  .addExample('DrilldownMenu/story/Default.example.tsx', 'Default')
  .addExample('DrilldownMenu/story/Dropdown.example.tsx', 'Dropdown')
