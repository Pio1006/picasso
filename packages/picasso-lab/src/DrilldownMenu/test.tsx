import React from 'react'
import { fireEvent, render, TestingPicasso } from '@toptal/picasso/test-utils'

import DrilldownMenu from './DrilldownMenu'

const TestDrilldownMenu = () => {
  const menuForItemB1 = (
    <DrilldownMenu data-testid='menu-b1'>
      <DrilldownMenu.Item data-testid='item-b1-1'>Item B1-1</DrilldownMenu.Item>
      <DrilldownMenu.Item data-testid='item-b1-2'>Item B1-2</DrilldownMenu.Item>
    </DrilldownMenu>
  )

  const menuForItemB2 = (
    <DrilldownMenu data-testid='menu-b2'>
      <DrilldownMenu.Item data-testid='item-b2-1'>Item B2-1</DrilldownMenu.Item>
      <DrilldownMenu.Item data-testid='item-b2-2'>Item B2-2</DrilldownMenu.Item>
    </DrilldownMenu>
  )

  const menuForItemB = (
    <DrilldownMenu data-testid='menu-b'>
      <DrilldownMenu.Item menu={menuForItemB1} data-testid='item-b1'>
        Item B1
      </DrilldownMenu.Item>
      <DrilldownMenu.Item menu={menuForItemB2} data-testid='item-b2'>
        Item B2
      </DrilldownMenu.Item>
    </DrilldownMenu>
  )

  return (
    <TestingPicasso>
      <DrilldownMenu data-testid='menu'>
        <DrilldownMenu.Item data-testid='item-a'>Item A</DrilldownMenu.Item>
        <DrilldownMenu.Item menu={menuForItemB} data-testid='item-b'>
          Item B
        </DrilldownMenu.Item>
      </DrilldownMenu>
    </TestingPicasso>
  )
}

describe('DrilldownMenu', () => {
  it('renders', () => {
    const { container } = render(<TestDrilldownMenu />)

    expect(container).toMatchSnapshot()
  })

  it('opens the first level menu on mouse enter', () => {
    const { container, getByTestId, queryByTestId } = render(
      <TestDrilldownMenu />
    )

    fireEvent.mouseEnter(getByTestId('item-b'))

    expect(queryByTestId('menu-b')).toBeInTheDocument()
    expect(queryByTestId('menu-b1')).not.toBeInTheDocument()
    expect(queryByTestId('menu-b2')).not.toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('opens the second level menu on mouse enter', () => {
    const { container, getByTestId, queryByTestId } = render(
      <TestDrilldownMenu />
    )

    fireEvent.mouseEnter(getByTestId('item-b'))
    fireEvent.mouseEnter(getByTestId('item-b1'))

    expect(queryByTestId('menu-b')).toBeInTheDocument()
    expect(queryByTestId('menu-b1')).toBeInTheDocument()
    expect(queryByTestId('menu-b2')).not.toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('closes menu on mouse leave', () => {
    const { container, getByTestId, queryByTestId } = render(
      <TestDrilldownMenu />
    )

    fireEvent.mouseEnter(getByTestId('item-b'))
    fireEvent.mouseLeave(getByTestId('menu-b'))

    expect(queryByTestId('menu')).toBeInTheDocument()
    expect(queryByTestId('menu-b')).not.toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('closes menu on click', () => {
    const { container, getByTestId, queryByTestId } = render(
      <TestDrilldownMenu />
    )

    fireEvent.mouseEnter(getByTestId('item-b'))
    fireEvent.click(document.body)

    expect(queryByTestId('menu')).toBeInTheDocument()
    expect(queryByTestId('menu-b')).not.toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })
})
