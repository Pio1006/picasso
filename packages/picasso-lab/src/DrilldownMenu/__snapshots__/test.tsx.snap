// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrilldownMenu closes menu on click 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="Picasso-root"
    >
      <ul
        class="MuiList-root PicassoMenuList-root"
        data-testid="menu"
        role="menu"
        tabindex="-1"
      >
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          data-testid="item-a"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item A
              </span>
            </div>
          </div>
        </li>
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          data-testid="item-b"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item B
              </span>
              <div
                class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
              >
                <svg
                  class="PicassoSvgChevronMinor16-root"
                  style="min-width: 16px; min-height: 16px;"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`DrilldownMenu opens the first level menu on click 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="Picasso-root"
    >
      <ul
        class="MuiList-root PicassoMenuList-root"
        data-testid="menu"
        role="menu"
        tabindex="-1"
      >
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          data-testid="item-a"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item A
              </span>
            </div>
          </div>
        </li>
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light Mui-selected PicassoMenuListItem-selected MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button Mui-selected"
          data-testid="item-b"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item B
              </span>
              <div
                class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
              >
                <svg
                  class="PicassoSvgChevronMinor16-root"
                  style="min-width: 16px; min-height: 16px;"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
      </ul>
      <div
        class="PicassoPopper-root"
        role="tooltip"
        style="position: absolute; top: 0px; left: 0px;"
        x-placement="right"
      >
        <div
          class="MuiPaper-root PicassoDrilldownMenuItem-content MuiPaper-elevation1"
        >
          <ul
            class="MuiList-root PicassoMenuList-root"
            data-testid="menu-b"
            role="menu"
            tabindex="-1"
          >
            <li
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
              data-testid="item-b1"
              role="menuitem"
              tabindex="0"
            >
              <div
                class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
              >
                <div
                  class="PicassoContainer-centerAlignItems PicassoContainer-flex"
                >
                  <span
                    class="PicassoMenuListItem-stringContent"
                  >
                    Item B1
                  </span>
                  <div
                    class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
                  >
                    <svg
                      class="PicassoSvgChevronMinor16-root"
                      style="min-width: 16px; min-height: 16px;"
                      viewBox="0 0 16 16"
                    >
                      <path
                        d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </li>
            <li
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
              data-testid="item-b2"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
              >
                <div
                  class="PicassoContainer-centerAlignItems PicassoContainer-flex"
                >
                  <span
                    class="PicassoMenuListItem-stringContent"
                  >
                    Item B2
                  </span>
                  <div
                    class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
                  >
                    <svg
                      class="PicassoSvgChevronMinor16-root"
                      style="min-width: 16px; min-height: 16px;"
                      viewBox="0 0 16 16"
                    >
                      <path
                        d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DrilldownMenu opens the second level menu on click 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="Picasso-root"
    >
      <ul
        class="MuiList-root PicassoMenuList-root"
        data-testid="menu"
        role="menu"
        tabindex="-1"
      >
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          data-testid="item-a"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item A
              </span>
            </div>
          </div>
        </li>
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light Mui-selected PicassoMenuListItem-selected MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button Mui-selected"
          data-testid="item-b"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item B
              </span>
              <div
                class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
              >
                <svg
                  class="PicassoSvgChevronMinor16-root"
                  style="min-width: 16px; min-height: 16px;"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
      </ul>
      <div
        class="PicassoPopper-root"
        role="tooltip"
        style="position: absolute; top: 0px; left: 0px;"
        x-placement="right"
      >
        <div
          class="MuiPaper-root PicassoDrilldownMenuItem-content MuiPaper-elevation1"
        >
          <ul
            class="MuiList-root PicassoMenuList-root"
            data-testid="menu-b"
            role="menu"
            tabindex="-1"
          >
            <li
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light Mui-selected PicassoMenuListItem-selected MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button Mui-selected"
              data-testid="item-b1"
              role="menuitem"
              tabindex="0"
            >
              <div
                class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
              >
                <div
                  class="PicassoContainer-centerAlignItems PicassoContainer-flex"
                >
                  <span
                    class="PicassoMenuListItem-stringContent"
                  >
                    Item B1
                  </span>
                  <div
                    class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
                  >
                    <svg
                      class="PicassoSvgChevronMinor16-root"
                      style="min-width: 16px; min-height: 16px;"
                      viewBox="0 0 16 16"
                    >
                      <path
                        d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </li>
            <li
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
              data-testid="item-b2"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
              >
                <div
                  class="PicassoContainer-centerAlignItems PicassoContainer-flex"
                >
                  <span
                    class="PicassoMenuListItem-stringContent"
                  >
                    Item B2
                  </span>
                  <div
                    class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
                  >
                    <svg
                      class="PicassoSvgChevronMinor16-root"
                      style="min-width: 16px; min-height: 16px;"
                      viewBox="0 0 16 16"
                    >
                      <path
                        d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="PicassoPopper-root"
        role="tooltip"
        style="position: absolute; top: 0px; left: 0px;"
        x-placement="right"
      >
        <div
          class="MuiPaper-root PicassoDrilldownMenuItem-content MuiPaper-elevation1"
        >
          <ul
            class="MuiList-root PicassoMenuList-root"
            data-testid="menu-b1"
            role="menu"
            tabindex="-1"
          >
            <li
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
              data-testid="item-b1-1"
              role="menuitem"
              tabindex="0"
            >
              <div
                class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
              >
                <div
                  class="PicassoContainer-centerAlignItems PicassoContainer-flex"
                >
                  <span
                    class="PicassoMenuListItem-stringContent"
                  >
                    Item B1-1
                  </span>
                </div>
              </div>
            </li>
            <li
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
              data-testid="item-b1-2"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
              >
                <div
                  class="PicassoContainer-centerAlignItems PicassoContainer-flex"
                >
                  <span
                    class="PicassoMenuListItem-stringContent"
                  >
                    Item B1-2
                  </span>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DrilldownMenu renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="Picasso-root"
    >
      <ul
        class="MuiList-root PicassoMenuList-root"
        data-testid="menu"
        role="menu"
        tabindex="-1"
      >
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          data-testid="item-a"
          role="menuitem"
          tabindex="0"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item A
              </span>
            </div>
          </div>
        </li>
        <li
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root PicassoMenuListItem-root PicassoMenuListItem-light MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
          data-testid="item-b"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoMenuListItem-content"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <span
                class="PicassoMenuListItem-stringContent"
              >
                Item B
              </span>
              <div
                class="PicassoContainer-leftxsmallMargin PicassoContainer-flex PicassoContainer-inline"
              >
                <svg
                  class="PicassoSvgChevronMinor16-root"
                  style="min-width: 16px; min-height: 16px;"
                  viewBox="0 0 16 16"
                >
                  <path
                    d="M5.997 3.29l4 4 .707.707-.707.707-4 4-.707-.707 4-4-4-4 .707-.707z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
